drafts/AENS.md

```diff
@@ -0,0 +1,414 @@
+and under what circumstances they are allowed to do so.
+
+The only available registrars for this AENS version will be hard-coded
+ones, which own the `.`, `.aet` and `.test` namespaces.
+
+The `.` namespace registrar is restricted and MUST NOT allow anyone to
+claim any names in its namespace.
+
+The `.aet` namespace registrar allows users to claim any name that is
+valid according to the validation rules for labels and names if they
+are the first to claim the name and pay the appropriate fee.
+
+The `.test` namespace registrar can be, as the name suggests, used for
+test purposes
+
+
+### Commitment Scheme
+
+Claiming a name requires a commitment scheme, which enables the claimant
+to commit to a name and after the commitment has been accepted into the 
+chain, reveal the name to finish the process.
+
+A commitment should be binding, i.e. the claimant cannot change
+the value they commited to withouth changing the actual commitment
+and hiding, so that a malicious miner learns nothing about the value
+the claimant has commited until they chose to reveal that value. This
+prevents malicious miners from front running, i.e. upon seeing a 
+claim transaction, including their own claim request for the same
+name instead of the original claimant's one.
+
+
+### Hashing
+
+We are going to adopt the `NameHash` as defined by the ENS in 
+[EIP-137](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm)
+using SHA256 instead.
+
+```
+def namehash(name):
+  if name == '':
+    return '\0' * 32
+  else:
+    label, _, remainder = name.partition('.')
+    return sha256(namehash(remainder) + sha256(label))
+```
+
+Names are generally only referred to in hashed form.
+
+### Protocol
+
+```
+                  expire
+      unclaimed <-------- revoked
+          | ^              ^^
+          | |              || expire
+          | | expire       ||
+pre-claim | |              ||  _
+          | |       revoke || | | transfer
+          v |              || | v
+     pre-claimed -------> claimed
+                  claim    | ^   
+                           | |
+                            -
+                          update
+```
+
+Note that `expire` is not an explicit message that is part
+of the protocol.
+
+
+#### Pre-claim
+
+Send a pre-claim transaction containing a hash commitment.
+
+```
+ ------------ ----
+| hash       | 32 |
+ ------------ ----
+```
+
+
+
+#### Claim
+
+```
+ ------------ ----
+| name       | 63 |
+ ------------ ----
+| fee        | 32 |
+ ------------ ----
+```
+
+Flow for a user:
+
+1. (optional) wait `n` blocks, s.t. that the block including the `pre-claim` cannot be reversed whp 
+2. send `claim` transaction to reveal name and pay the associated fee
+
+If the time delta of 1 and 2 is bigger than e.g. 48 hours then the 
+`pre-claim` expires again.
+
+The `claim` transaction MUST be signed by the same private key as the 
+`pre-claim` transaction.
+
```

Comment:
What happens if the name is already claimed?
Will the claim transaction be considered invalid an not included, hence no fee, or will it be consider just failing and still costing the fee?

Reply:
If a hash/name is already claimed then the transaction can be considered invalid. Ideally a wallet would warn a user trying to claim such a name.